Source/Dafny/Verifier/Translator.cs

```diff
@@ -7744,15 +7744,15 @@ AND  Apply(f,h0,bxs) == Apply(f,h0,bxs)
           return predef.ArrayLength;
         } else if (f.EnclosingClass == null && f.Name == "Floor") {
           return predef.RealFloor;
-        } else if (f is SpecialField && (f.Name == "Keys" || f.Name == "Values" || f.Name == "Items")) {
-          Contract.Assert(f.Type is SetType);
-          var setType = (SetType)f.Type;
-          if (f.Name == "Keys") {
-            return setType.Finite ? predef.MapDomain : predef.IMapDomain;
-          } else if (f.Name == "Values") {
-            return setType.Finite ? predef.MapValues : predef.IMapValues;
-          } else {
-            return setType.Finite ? predef.MapItems : predef.IMapItems;
+        } else if (f is SpecialField && !(f is DatatypeDestructor)) {
+          if (f.Name == "Keys" || f.Name == "Values" || f.Name == "Items") {
+            Contract.Assert(f.Type is SetType);
+            var setType = (SetType)f.Type;
+            return f.Name switch {
+              "Keys" => setType.Finite ? predef.MapDomain : predef.IMapDomain,
+              "Values" => setType.Finite ? predef.MapValues : predef.IMapValues,
+              _ => setType.Finite ? predef.MapItems : predef.IMapItems
+            };
           }
         } else if (f is SpecialField && f.Name == "IsLimit") {
```

Comment:
Instead of checking `f.IsSpecialField` repeatedly we can move this case and the following ones into the switch above